content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md

```diff
@@ -0,0 +1,558 @@
+	//
+	// The reason there's more than one stack is because we might be unmarshaling
+	// a single JSON value into multiple GraphQL fragments or embedded structs, so
+	// we keep track of them all.
+	vs [][]reflect.Value
+}
+```
+
+We need to modify `decode` to create additional stacks whenever we encounter an embedded struct or a struct with `graphql:"... on Type"` field tag, do some additional bookkeeping to manage multiple stacks of values, check for additional error conditions if our stacks run empty. Aside from that, the same algorithm continues to work.
+
+I think getting the data structure to contain just enough information to resolve the task was the most challenging part of getting this to work. Once it's there, the rest of the algorithm details fall into place.
+
+If you'd like to learn even more of the low-level details of the implementation, I invite you to look at the [source code](https://github.com/shurcooL/graphql/blob/master/internal/jsonutil/graphql.go) of package [`github.com/shurcooL/graphql/internal/jsonutil`](https://godoc.org/github.com/shurcooL/graphql/internal/jsonutil). It should be easy to read after this post.
+
+Payoff
+------
+
+Let's quickly revisit our original GraphQL unions example that wasn't working with standard `encoding/json` unmarshaler. When we replace `json.UnmarshalJSON` with `jsonutil.UnmarshalGraphQL`, the `Name` field gets populated! That's good news, it means we didn't do all that work for nothing.
+
+See proof on [playground](https://play.golang.org/p/Xfu2mqxZ5m).
+
+`jsonutil.UnmarshalGraphQL` also takes `graphql` struct field tags into account when unmarshaling, and doesn't get mislead by `json` field tags. If there are additional GraphQL-specific tweaks to unmarshaling behavior that need to be applied in the future, they'll be easy to apply. Best part is we're reusing the rigorous JSON tokenizer of `encoding/json` and its public API, so no need to deal with maintaining a fork.
+
+Conclusion
+----------
+
+If you got this far, thanks for following along with me on this journey! I hope you enjoyed it and/or learned something new.
+
+It has been a lot of fun implementing the GraphQL client library for Go, and trying to make the best [API design decisions](https://github.com/shurcooL/githubql/issues?q=label%3A%22API+decision%22). I enjoyed using the tools that Go gives me to tackle this task. Even after using Go for 4 years, I'm still finding Go to be the absolutely most fun programming language to use, and feeling same joy I did back when I was just starting out!
+
+I think GraphQL is an exciting new technology. Its strongly typed nature is a great fit for Go. APIs that are created with it can be a pleasure to use. Keep in mind that GraphQL shines most when you're able to replace multiple REST API calls with a single carefully crafted GraphQL query. This requires high quality and completeness of the GraphQL schema, so not all GraphQL APIs are made equal.
```

Comment:
s/I think GraphQL is/GraphQL is. We know who thinks it is exciting, you're writing the post!